packages/block-editor/src/hooks/utils.js

```diff
@@ -33,7 +33,7 @@ export const cleanEmptyObject = ( object ) => {
 	const cleanedNestedObjects = Object.fromEntries(
 		Object.entries( object )
 			.map( ( [ key, value ] ) => [ key, cleanEmptyObject( value ) ] )
-			.filter( ( [ , value ] ) => Boolean( value ) )
+			.filter( ( [ , value ] ) => value !== undefined )
```

Comment:
The concerning bit is that we don't have a good way to check all instances if they really mean to filter `undefined` values only. 

First off, `undefined` might be a valid value here or there, and we might not want to clean that up. But that's fine to ignore since it was handled before and it is being handled with this new version, too.

On the other hand, though, there are some expectations that the cleaning function will also clean empty strings. You can see that there are a few failing E2E tests that confirm that problem. 

So, I'd suggest reverting that change, keeping the `Boolean()` for now, and exploring changing it to a separate PR. 

Maybe the clean function can have an argument, controlling whether to filter boolean values or undefined values, although that may seem a bit too over-engineered. 

In any case - exploring it separately would make the most sense, and with that change reverted, the PR still has a lot of value in removing duplicate code and introducing a shared function that is already necessary for multiple areas of the code.

Reply:
> The concerning bit is that we don't have a good way to check all instances if they really mean to filter undefined values only.

This gave me pause for concern as well over on https://github.com/WordPress/gutenberg/pull/48917/ where I was needing to allow zero values which get filtered by the `Boolean()` call.

> Maybe the clean function can have an argument, controlling whether to filter boolean values or undefined values, although that may seem a bit too over-engineered.

https://github.com/WordPress/gutenberg/pull/48917 followed a similar approach adding an `options` argument and a flag to `allowZeros`. We might be able to get away with the simpler option to decide between boolean or undefined checks.

> In any case - exploring it separately would make the most sense, and with that change reverted, the PR still has a lot of value

+1 Agreed.

Reply:
> The concerning bit is that we don't have a good way to check all instances if they really mean to filter undefined values only.

I'm counting 22 places where this function is used. It's not a public API, so extenders don't need to be taken into account. It's not impossible to manually test all those instances and make sure they behave the same as in trunk 😅 

I've been testing adding/removing/resetting values in all the tools panel controls and everything seems to work as expected, plus the e2e tests are passing now, so I'm guessing that the introduction of `normalizeFalsyValue` has solved most of the problems.